src/core/manager.c

```diff
@@ -1729,7 +1729,8 @@ static void invoke_sigchld_event(Manager *m, Unit *u, const siginfo_t *si) {
         unit_unwatch_pid(u, si->si_pid);
 
         if (UNIT_VTABLE(u)->sigchld_event) {
-                if (set_size(u->pids) <= 1 || iteration != u->sigchldgen) {
+                if (set_size(u->pids) <= 1 || iteration != u->sigchldgen ||
+                    (u->type == UNIT_SERVICE && (si->si_pid == SERVICE(u)->main_pid || si->si_pid == SERVICE(u)->control_pid))) {
```

Comment:
We should really avoid "reaching over" into the specific unit type here, from generic code. And thankfully, we don't have to: there's unit_control_pid() and unit_main_pid() which you can call here, generically without casting the unit to any specific type. Please use that!